store/src/commonMain/kotlin/com/dropbox/external/store5/impl/ShareableMarket.kt

```diff
@@ -0,0 +1,431 @@
+@file:Suppress("UNCHECKED_CAST")
+
+package com.dropbox.external.store5.impl
+
+import com.dropbox.external.store5.ConflictResolver
+import com.dropbox.external.store5.Fetcher
+import com.dropbox.external.store5.Market
+import com.dropbox.external.store5.MarketResponse
+import com.dropbox.external.store5.OnMarketCompletion
+import com.dropbox.external.store5.OnRemoteCompletion
+import com.dropbox.external.store5.Reader
+import com.dropbox.external.store5.RemoteResult
+import com.dropbox.external.store5.Store
+import com.dropbox.external.store5.Updater
+import com.dropbox.external.store5.Write
+import com.dropbox.external.store5.Writer
+import com.dropbox.external.store5.definition.Converter
+import com.dropbox.external.store5.definition.PostRequest
+import com.dropbox.external.store5.operator.shareableMutableMapOf
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.async
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.MutableSharedFlow
+import kotlinx.coroutines.flow.lastOrNull
+import kotlinx.coroutines.launch
+import kotlinx.datetime.Clock
+
+
+typealias AnyReadCompletionsQueue = MutableList<OnMarketCompletion<*>>
+typealias AnyWriteRequestQueue<Key> = ArrayDeque<Updater<Key, *, *>>
+typealias SomeWriteRequestQueue<Key, Input> = ArrayDeque<Updater<Key, Input, *>>
+typealias AnyBroadcast = MutableSharedFlow<MarketResponse<*>>
+typealias SomeBroadcast<T> = MutableSharedFlow<MarketResponse<T>>
+
+
+/**
+ * Thread-safe [Market] implementation.
+ * @param stores List of [Store]. Order matters! [ShareableMarket] executes [read], [write], and [delete] operations iteratively.
+ * @param conflictResolver Implementation of [ConflictResolver]. Used in [read] to eagerly resolve conflicts and in [write] to persist write request failures.
+ * @property readCompletions Thread-safe mapping of Key to [AnyReadCompletionsQueue]. Queue size is checked before handling a market read request. All completion handlers in queue are processed on market read response.
+ * @property writeRequests Thread-safe mapping of Key to [AnyWriteRequestQueue], an alias of an [Updater] queue. All requests are put in the queue. In [tryPost], we get the most recent request from the queue. Then we get the last value from [SomeBroadcast]. On response, write requests are handled based on their time of creation. On success, we invoke [resetWriteRequestQueue]. On failure, we use [ConflictResolver.setLastFailedWriteTime]. This map is only saved in memory. However, last failed write time and last local value will persist as long as the [Market] contains a persistent [Store].
+ * @property broadcasts Thread-safe mapping of Key to [AnyBroadcast], an alias of a Mutable Shared Flow of [MarketResponse]. Callers of [read] receive [SomeBroadcast], which is the typed equivalent of [AnyBroadcast].
+ */
+class ShareableMarket<Key : Any> internal constructor(
+    private val scope: CoroutineScope,
```

Comment:
Question on `scope`. When do we want to use it for? Do we want to use to trigger emissions? Or are we saying that all methods will dispatched to this scope? (In which case we can remove the`suspend` functions?). 

I think ideally, we would only need a `scope` if there's something that we need to launch or `listen` to. But I don't think we need to do that here?

Reply:
Ah yeah will fix 👍🏽